segmentstore/server/src/test/java/io/pravega/segmentstore/server/DataRecoveryTestUtils.java

```diff
@@ -0,0 +1,188 @@
+            log.info("Recovery started for container = {}", containerId);
+            ContainerTableExtension ext = container.getExtension(ContainerTableExtension.class);
+            AsyncIterator<IteratorItem<TableKey>> it = ext.keyIterator(getMetadataSegmentName(containerId),
+                    IteratorArgs.builder().fetchTimeout(TIMEOUT).build()).join();
+
+            // Add all segments present in the container metadata in a set.
+            Set<TableKey> segmentsInMD = new HashSet<>();
+            it.forEachRemaining(k -> segmentsInMD.addAll(k.getEntries()), EXECUTOR_SERVICE).join();
+
+            for (SegmentProperties segment : segments) {
+                long len = segment.getLength();
+                boolean isSealed = segment.isSealed();
+                String segmentName = segment.getName();
+
+                /*
+                    1. segment exists in both metadata and storage, re-create it
+                    2. segment only in metadata, delete
+                    3. segment only in storage, re-create it
+                 */
+                segmentsInMD.remove(TableKey.unversioned(getTableKey(segmentName)));
+                container.getStreamSegmentInfo(segment.getName(), TIMEOUT)
+                        .thenAccept(e -> {
+                            container.createStreamSegment(segmentName, len, isSealed)
+                                    .exceptionally(ex -> {
+                                        log.error("Exception occurred while creating segment", ex);
+                                        return null;
+                                    }).join();
+                        })
+                        .exceptionally(e -> {
+                            log.error("Got an exception on getStreamSegmentInfo", e);
+                            if (Exceptions.unwrap(e) instanceof StreamSegmentNotExistsException) {
+                                container.createStreamSegment(segmentName, len, isSealed)
+                                        .exceptionally(ex -> {
+                                            log.error("Exception occurred while creating segment", ex);
+                                            return null;
+                                        }).join();
+                            }
+                            return null;
+                        }).join();
+            }
+            for (TableKey k : segmentsInMD) {
+                String segmentName = k.getKey().toString();
+                log.info("Deleting segment : {} as it is not in storage", segmentName);
+                try {
+                    container.deleteStreamSegment(segmentName, TIMEOUT).join();
+                } catch (Throwable e) {
+                    log.error("Error while deleting the segment = {}", segmentName);
+                }
+            }
+            log.info("Recovery done for container = {}", containerId);
+        }
+    }
+
+    public static ArrayView getTableKey(String segmentName) {
```

Comment:
You won't need this anymore if you implement my suggestion above.

Reply:
Yes. Done.